docs/source/build_python_model.rst

```diff
@@ -1,4 +1,104 @@
+      act = sg.nn.convolution(
+          act, conv_tensor, stride=[1, 1], padding="same", activation="relu")
+      # After the max pooling, shape of act: [1, 32, 14, 14].
+      act = sg.nn.max_pool(act, pool_size=[2, 2], stride=[2, 2])
+      # After the matrix multiply, shape of act: [1, 10].
+      act = sg.nn.mat_mul(act, fc_tensor)
+
+As we create the first operator for the model, we need to first prepare an input
+tensor and weight tensors that are used by the operator. In the example, we
+create an input tensor `input_tensor` using the :class:`smaug.Tensor` API. Here,
+we specify the `data_layout` as `sg.NHWC`, which stands for a 4D tensor shape
+with the channel-major layout. We also specify the `tensor_data` parameter with
+a randomly generated NumPy array, with a shape of [1, 28, 28, 1]. However, the
+user can use the real weights extracted from a pretrained model. Likewise, we
+create two weight tensors that will be used by a convolution operator and a
+matrix multiply operator, respectively.
+
+The `sg.input_data` creates a data operator that simply creates a reference
+tensor `act` to its input data `input_tensor`. Then, `act` is fed to a
+convolution operator that also takes `conv_tensor` as its filter input. With
+more details provided in :func:`smaug.nn.convolution`, it computes a 3D
+convolution given the 4D input and filter tensors, and we use 1x1 strides,
+the `same` padding and a ReLU activation fused with the convolution operation.
+The output of it then goes through a max pooling operator with a 2x2 filter
+size, which in turn fans its output into the last matrix multiply operator. Note
+that since the output of the max pooling operator is a 4D tensor while
+:func:`smaug.nn.mat_mul` expects a 2D input tensor, SMAUG will automatically
+add a layout transformation operator :func:`smaug.tensor.reorder` in between to
+make the data layout format compatible. Thus, the 4D tensor of shape
+[1, 32, 14, 14] will be flattened into a 2D tensor of shape [1, 6272] before
+running the matrix multiply. Similarly, SMAUG will also perform the NHWC to NCHW
+layout transformation or vice versa as per the expected layout format of the
+backend.
+
+After finishing adding operators to the model, we can now take a look at the
+summary of the model using the :func:`smaug.Graph.print_summary` API, which
+prints model-level information and operator-specific properties.
+
+.. code-block:: python
+
+    graph.print_summary()
+
+Finally, we can export the model files using the
+:func:`smaug.Graph.write_graph` API.
+
+.. code-block:: python
+
+    graph.write_graph()
+
+This gives us two files named `my_model_topo.pbtxt` and `my_model_params.pb`,
+where the former stores all the model information except for the parameters,
+which are stored in the latter. This separation is helpful for us to quickly
+check things in the human readable topology file while still compressing as
+much as possible the oftentimes large paramaters.
```

Comment:
Finally, you should link this up with the C++ side by showing how to run the model. You can leave the details to the C++ API which already explains this.

Reply:
Right. Linked this up with the C++ side.